File: examples/cpp/source/quality_metrics/lin_reg_metrics_dense_batch.cpp
/* file: lin_reg_metrics_dense_batch.cpp */
/*******************************************************************************
* Copyright 2014-2016 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

/*
!  Content:
!    C++ example of linear regression quality metrics in batch processing mode.
!
!    The program trains the linear regression model on a training
!    datasetFileName with normal equation and QR methods and computes quality
!    metrics for the models.
!******************************************************************************/

/**
* <a name="DAAL-EXAMPLE-CPP-LIN_REG_QUALITY_METRIC_SET_BATCH"></a>
* \example lin_reg_metrics_dense_batch.cpp
*/

#include "daal.h"
#include "service.h"
#include "algorithms/linear_regression/linear_regression_quality_metric_set_batch.h"
#include "algorithms/linear_regression/linear_regression_quality_metric_set_types.h"
#include "algorithms/linear_regression/linear_regression_single_beta_batch.h"
#include "algorithms/linear_regression/linear_regression_single_beta_types.h"
#include "algorithms/linear_regression/linear_regression_group_of_betas_types.h"
#include "algorithms/linear_regression/linear_regression_group_of_betas_batch.h"
#include <strstream>

using namespace std;
using namespace daal;
using namespace daal::services;
using namespace daal::algorithms::linear_regression;
using namespace daal::algorithms::linear_regression::quality_metric;

/* Input data set parameters */
string trainDatasetFileName = "../data/batch/linear_regression_train.csv";
string testDatasetFileName = "../data/batch/linear_regression_test.csv";

const size_t nFeatures = 10;
const size_t nDependentVariables = 2;

template <class TrainingAlgorithm>
void trainModel(TrainingAlgorithm& algo);
void testModelQuality();
void printResults();

services::SharedPtr<training::Result> trainingResult;
services::SharedPtr<quality_metric_set::ResultCollection> qmsResult;
NumericTablePtr trainData;
NumericTablePtr trainDependentVariables;

int main(int argc, char *argv[])
{
    checkArguments(argc, argv, 2, &trainDatasetFileName, &testDatasetFileName);

    /* Initialize FileDataSource<CSVFeatureManager> to retrieve the input data from a .csv file */
    FileDataSource<CSVFeatureManager> trainDataSource(trainDatasetFileName,
        DataSource::notAllocateNumericTable,
        DataSource::doDictionaryFromContext);

    /* Create Numeric Tables for training data and dependent variables */
    trainData = NumericTablePtr(new HomogenNumericTable<double>(nFeatures, 0, NumericTable::notAllocate));
    trainDependentVariables = NumericTablePtr(new HomogenNumericTable<double>(nDependentVariables, 0, NumericTable::notAllocate));
    NumericTablePtr mergedData(new MergedNumericTable(trainData, trainDependentVariables));

    /* Retrieve the data from input file */
    trainDataSource.loadDataBlock(mergedData.get());

    for(size_t i = 0; i < 2; ++i)
    {
        if(i == 0)
        {
            /* Create an algorithm object to train the multiple linear regression model with normal equation method */
            std::cout << "Train model with normal equation algorithm." << std::endl;
            training::Batch<> algorithm;
            trainModel(algorithm);
        }
        else
        {
            /* Create an algorithm object to train the multiple linear regression model with QR method */
            std::cout << "Train model with QR algorithm." << std::endl;
            training::Batch<double, training::qrDense> algorithm;
            trainModel(algorithm);
        }
        testModelQuality();
        printResults();
    }
    return 0;
}

template <class TrainingAlgorithm>
void trainModel(TrainingAlgorithm& algorithm)
{
    /* Pass a training data set and dependent values to the algorithm */
    algorithm.input.set(training::data, trainData);
    algorithm.input.set(training::dependentVariables, trainDependentVariables);

    /* Build the multiple linear regression model */
    algorithm.compute();

    /* Retrieve the algorithm results */
    trainingResult = algorithm.getResult();
    printNumericTable(trainingResult->get(training::model)->getBeta(), "Linear Regression coefficients:");
}

NumericTablePtr predictResults(NumericTablePtr& data)
{
    /* Create an algorithm object to predict values of multiple linear regression */
    prediction::Batch<> algorithm;

    /* Pass a testing data set and the trained model to the algorithm */
    algorithm.input.set(prediction::data, data);
    algorithm.input.set(prediction::model, trainingResult->get(training::model));

    /* Predict values of multiple linear regression */
    algorithm.compute();

    /* Retrieve the algorithm results */
    services::SharedPtr<prediction::Result> predictionResult = algorithm.getResult();
    return predictionResult->get(prediction::prediction);
}

NumericTablePtr predictReducedModelResults(NumericTablePtr& data)
{
    services::SharedPtr<Model> model = trainingResult->get(training::model);

    NumericTablePtr betas = model->getBeta();
    const size_t nBetas = model->getNumberOfBetas();

    /* Set beta coefficients #2 and #10 to zero */
    size_t j1 = 2;
    size_t j2 = 10;
    double* savedBeta = new double[nBetas * nDependentVariables];
    {
        BlockDescriptor<> block;
        betas->getBlockOfRows(0, nDependentVariables, readWrite, block);
        double* pBeta = block.getBlockPtr();
        for(size_t i = 0; i < nDependentVariables; ++i)
        {
            savedBeta[nDependentVariables*i + j1] = pBeta[nDependentVariables*i + j1];
            savedBeta[nDependentVariables*i + j2] = pBeta[nDependentVariables*i + j2];
            pBeta[nDependentVariables*i + j1] = 0;
            pBeta[nDependentVariables*i + j2] = 0;
        }
        betas->releaseBlockOfRows(block);
    }

    /* Predict with reduced model */
    NumericTablePtr predictedResults = predictResults(trainData);

    /* Restore the coefficients */
    {
        BlockDescriptor<> block;
        betas->getBlockOfRows(0, nDependentVariables, readWrite, block);
        double* pBeta = block.getBlockPtr();
        for(size_t i = 0; i < nDependentVariables; ++i)
        {
            pBeta[nDependentVariables*i + j1] = savedBeta[nDependentVariables*i + j1];
            pBeta[nDependentVariables*i + j2] = savedBeta[nDependentVariables*i + j2];
        }
        betas->releaseBlockOfRows(block);
    }
    delete[] savedBeta;
    return predictedResults;
}

void testModelQuality()
{
    /* Predict results with the full model */
    NumericTablePtr predictedResults = predictResults(trainData);
    printNumericTable(trainDependentVariables, "Expected responses (first 20 rows):", 20);
    printNumericTable(predictedResults, "Predicted responses (first 20 rows):", 20);

    services::SharedPtr<Model> model = trainingResult->get(training::model);

    /* Predict results with the reduced model */
    NumericTablePtr predictedReducedModelResults = predictReducedModelResults(trainData);
    printNumericTable(predictedReducedModelResults, "Responses predicted with reduced model (first 20 rows):", 20);

    /* Create a quality metric set object to compute quality metrics of the linear regression algorithm */
    const size_t nBetaReducedModel = model->getNumberOfBetas() - 2;
    quality_metric_set::Batch qualityMetricSet(model->getNumberOfBetas(), nBetaReducedModel);

    /* Set input for a single beta metrics algorithm */
    services::SharedPtr<algorithms::Input> algInput =
        qualityMetricSet.getInputDataCollection()->getInput(quality_metric_set::singleBeta);
    single_beta::InputPtr singleBeta = single_beta::Input::cast(algInput);

    singleBeta->set(single_beta::expectedResponses, trainDependentVariables);
    singleBeta->set(single_beta::predictedResponses, predictedResults);
    singleBeta->set(single_beta::model, model);

    /* Set input for a group of betas metrics algorithm */
    algInput = qualityMetricSet.getInputDataCollection()->getInput(quality_metric_set::groupOfBetas);
    group_of_betas::InputPtr groupOfBetas = group_of_betas::Input::cast(algInput);
    groupOfBetas->set(group_of_betas::expectedResponses, trainDependentVariables);
    groupOfBetas->set(group_of_betas::predictedResponses, predictedResults);
    groupOfBetas->set(group_of_betas::predictedReducedModelResponses, predictedReducedModelResults);

    /* Compute quality metrics */
    qualityMetricSet.compute();

    /* Retrieve the quality metrics */
    qmsResult = qualityMetricSet.getResultCollection();
}

void printResults()
{
    /* Print the quality metrics for a single beta */
    std::cout << "Quality metrics for a single beta" << std::endl;
    {
        single_beta::ResultPtr result = single_beta::Result::cast(qmsResult->getResult(quality_metric_set::singleBeta));
        if(!result)
            return;
        printNumericTable(result->get(single_beta::rms), "Root means square errors for each response (dependent variable):");
        printNumericTable(result->get(single_beta::variance), "Variance for each response (dependent variable):");
        printNumericTable(result->get(single_beta::zScore), "Z-score statistics:");
        printNumericTable(result->get(single_beta::confidenceIntervals), "Confidence intervals for each beta coefficient:");
        printNumericTable(result->get(single_beta::inverseOfXtX), "Inverse(Xt * X) matrix:");

        data_management::DataCollectionPtr coll = result->get(single_beta::betaCovariances);
        for(size_t i = 0; i < coll->size(); ++i)
        {
            std::ostringstream str;
            str << "Variance-covariance matrix for betas of " << i << "-th response" << std::endl;
            NumericTablePtr betaCov = data_management::NumericTable::cast((*coll)[i]);
            printNumericTable(betaCov, str.str().c_str());
        }
    }
    /* Print quality metrics for a group of betas */
    std::cout << "Quality metrics for a group of betas" << std::endl;
    {
        group_of_betas::ResultPtr result = group_of_betas::Result::cast(qmsResult->getResult(quality_metric_set::groupOfBetas));
        if(!result)
            return;

        printNumericTable(result->get(group_of_betas::expectedMeans), "Means of expected responses for each dependent variable:", 0, 0, 20);
        printNumericTable(result->get(group_of_betas::expectedVariance), "Variance of expected responses for each dependent variable:", 0, 0, 20);
        printNumericTable(result->get(group_of_betas::regSS), "Regression sum of squares of expected responses:", 0, 0, 20);
        printNumericTable(result->get(group_of_betas::resSS), "Sum of squares of residuals for each dependent variable:", 0, 0, 20);
        printNumericTable(result->get(group_of_betas::tSS), "Determination coefficient for each dependent variable:", 0, 0, 20);
        printNumericTable(result->get(group_of_betas::fStatistics), "F-statistics for each dependent variable:", 0, 0, 20);
    }
}

File: examples/python/source/quality_metrics/lin_reg_metrics_dense_batch.py
# file: lin_reg_metrics_dense_batch.py
#===============================================================================
# Copyright 2014-2019 Intel Corporation.
#
# This software and the related documents are Intel copyrighted  materials,  and
# your use of  them is  governed by the  express license  under which  they were
# provided to you (License).  Unless the License provides otherwise, you may not
# use, modify, copy, publish, distribute,  disclose or transmit this software or
# the related documents without Intel's prior written permission.
#
# This software and the related documents  are provided as  is,  with no express
# or implied  warranties,  other  than those  that are  expressly stated  in the
# License.
#===============================================================================

## <a name="DAAL-EXAMPLE-PY-LIN_REG_METRICS_DENSE_BATCH"></a>
## \example lin_reg_metrics_dense_batch.py

import os
import sys

import daal.algorithms.linear_regression as linear_regression
import daal.algorithms.linear_regression.quality_metric_set as quality_metric_set
from daal.algorithms.linear_regression import training, prediction
from daal.algorithms.linear_regression.quality_metric import single_beta, group_of_betas
from daal.data_management import (
    DataSourceIface, FileDataSource, HomogenNumericTable, MergedNumericTable,
    NumericTableIface, BlockDescriptor, readWrite
)

utils_folder = os.path.realpath(os.path.abspath(os.path.dirname(os.path.dirname(__file__))))
if utils_folder not in sys.path:
    sys.path.insert(0, utils_folder)
from utils import printNumericTable

trainDatasetFileName = os.path.join('..', 'data', 'batch', 'linear_regression_train.csv')

nFeatures = 10
nDependentVariables = 2

trainingResult = None
# predictionResult = None
qmsResult = None
trainData = None
trainDependentVariables = None

def trainModel(algorithm):
    global trainingResult, trainData, trainDependentVariables

    # Pass a training data set and dependent values to the algorithm
    algorithm.input.set(training.data, trainData)
    algorithm.input.set(training.dependentVariables, trainDependentVariables)

    # Build the multiple linear regression model and retrieve the algorithm results
    trainingResult = algorithm.compute()
    printNumericTable(trainingResult.get(training.model).getBeta(), "Linear Regression coefficients:")

def predictResults(trainData):
    # Create an algorithm object to predict values of multiple linear regression
    algorithm = prediction.Batch()

    # Pass a testing data set and the trained model to the algorithm
    algorithm.input.setTable(prediction.data, trainData)
    algorithm.input.setModel(prediction.model, trainingResult.get(training.model))

    # Predict values of multiple linear regression and retrieve the algorithm results
    predictionResult = algorithm.compute()
    return predictionResult.get(prediction.prediction)

def predictReducedModelResults(trainData):
    model = trainingResult.get(training.model)

    betas = model.getBeta()
    nBetas = model.getNumberOfBetas()

    j1 = 2
    j2 = 10
    savedBeta = [[None] * nBetas for _ in range(nDependentVariables)]

    block = BlockDescriptor()
    betas.getBlockOfRows(0, nDependentVariables, readWrite, block)
    pBeta = block.getArray()

    for i in range(0, nDependentVariables):
        savedBeta[i][j1] = pBeta[i][j1]
        savedBeta[i][j2] = pBeta[i][j2]
        pBeta[i][j1] = 0
        pBeta[i][j2] = 0
    betas.releaseBlockOfRows(block)

    predictedResults = predictResults(trainData)

    block = BlockDescriptor()
    betas.getBlockOfRows(0, nDependentVariables, readWrite, block)
    pBeta = block.getArray()

    for i in range(0, nDependentVariables):
        pBeta[i][j1] = savedBeta[i][j1]
        pBeta[i][j2] = savedBeta[i][j2]
    betas.releaseBlockOfRows(block)
    return predictedResults

def testModelQuality():
    global trainingResult, qmsResult

    predictedResults = predictResults(trainData)
    printNumericTable(trainDependentVariables, "Expected responses (first 20 rows):", 20)
    printNumericTable(predictedResults, "Predicted responses (first 20 rows):", 20)

    model = trainingResult.get(linear_regression.training.model)
    predictedReducedModelResults = predictReducedModelResults(trainData)
    printNumericTable(predictedReducedModelResults, "Responses predicted with reduced model (first 20 rows):", 20)

    # Create a quality metric set object to compute quality metrics of the linear regression algorithm
    nBetaReducedModel = model.getNumberOfBetas() - 2
    qualityMetricSet = quality_metric_set.Batch(model.getNumberOfBetas(), nBetaReducedModel)
    singleBeta = single_beta.Input.downCast(qualityMetricSet.getInputDataCollection().getInput(quality_metric_set.singleBeta))
    singleBeta.setDataInput(single_beta.expectedResponses, trainDependentVariables)
    singleBeta.setDataInput(single_beta.predictedResponses, predictedResults)
    singleBeta.setModelInput(single_beta.model, model)

    # Set input for a group of betas metrics algorithm
    groupOfBetas = group_of_betas.Input.downCast(qualityMetricSet.getInputDataCollection().getInput(quality_metric_set.groupOfBetas))
    groupOfBetas.set(group_of_betas.expectedResponses, trainDependentVariables)
    groupOfBetas.set(group_of_betas.predictedResponses, predictedResults)
    groupOfBetas.set(group_of_betas.predictedReducedModelResponses, predictedReducedModelResults)

    # Compute quality metrics
    qualityMetricSet.compute()

    # Retrieve the quality metrics
    qmsResult = qualityMetricSet.getResultCollection()

def printResults():
    # Print the quality metrics for a single beta
    print ("Quality metrics for a single beta")
    result = single_beta.Result.downCast(qmsResult.getResult(quality_metric_set.singleBeta))
    printNumericTable(result.getResult(single_beta.rms), "Root means square errors for each response (dependent variable):")
    printNumericTable(result.getResult(single_beta.variance), "Variance for each response (dependent variable):")
    printNumericTable(result.getResult(single_beta.zScore), "Z-score statistics:")
    printNumericTable(result.getResult(single_beta.confidenceIntervals), "Confidence intervals for each beta coefficient:")
    printNumericTable(result.getResult(single_beta.inverseOfXtX), "Inverse(Xt * X) matrix:")

    coll = result.getResultDataCollection(single_beta.betaCovariances)
    for i in range(0, coll.size()):
        message = "Variance-covariance matrix for betas of " + str(i) + "-th response\n"
        betaCov = result.get(single_beta.betaCovariances, i)
        printNumericTable(betaCov, message)

    # Print quality metrics for a group of betas
    print ("Quality metrics for a group of betas")
    result = group_of_betas.Result.downCast(qmsResult.getResult(quality_metric_set.groupOfBetas))

    printNumericTable(result.get(group_of_betas.expectedMeans), "Means of expected responses for each dependent variable:", 0, 0, 20)
    printNumericTable(result.get(group_of_betas.expectedVariance), "Variance of expected responses for each dependent variable:", 0, 0, 20)
    printNumericTable(result.get(group_of_betas.regSS), "Regression sum of squares of expected responses:", 0, 0, 20)
    printNumericTable(result.get(group_of_betas.resSS), "Sum of squares of residuals for each dependent variable:", 0, 0, 20)
    printNumericTable(result.get(group_of_betas.tSS), "Total sum of squares for each dependent variable:", 0, 0, 20)
    printNumericTable(result.get(group_of_betas.determinationCoeff), "Determination coefficient for each dependent variable:", 0, 0, 20)
    printNumericTable(result.get(group_of_betas.fStatistics), "F-statistics for each dependent variable:", 0, 0, 20)

if __name__ == "__main__":

    # Initialize FileDataSource<CSVFeatureManager> to retrieve the input data from a .csv file
    dataSource = FileDataSource(trainDatasetFileName,
                                DataSourceIface.notAllocateNumericTable,
                                DataSourceIface.doDictionaryFromContext)

    # Create Numeric Tables for data and values for dependent variable
    trainData = HomogenNumericTable(nFeatures, 0, NumericTableIface.doNotAllocate)
    trainDependentVariables = HomogenNumericTable(nDependentVariables, 0, NumericTableIface.doNotAllocate)
    mergedData = MergedNumericTable(trainData, trainDependentVariables)

    # Retrieve the data from the input file
    dataSource.loadDataBlock(mergedData)

    for i in range(0, 2):
        if i == 0:
            print ("Train model with normal equation algorithm.")
            algorithm = training.Batch()
            trainModel(algorithm)
        else:
            print ("Train model with QR algorithm.")
            algorithm = training.Batch(method=training.qrDense)
            trainModel(algorithm)
        testModelQuality()
        printResults()
